### PR TITLE
Tc/schedule templates

### PR DESCRIPTION
**Pull request by CedThen**

### CODE REVIEW

File: server/utilities/helperFunctions.ts [hunk 1]
```diff
@@ -329,11 +330,25 @@ async function retrieveTaskData(eventId, role) {
     return {
       taskId: task.id,
       date: task.date,
-      firstName: task.user.firstName,
-      lastName: task.user.lastName,
+      firstName: task.user?.firstName || '',
+      lastName: task.user?.lastName || '',
       userId: task.userId,
```
Comment (CedThen) [inline server/utilities/helperFunctions.ts:335]: handles if task is blank and does not have a userId

File: src/components/shared/Alert.tsx [hunk 2]
```diff
@@ -2,11 +2,11 @@ import React, { useEffect } from 'react';
...
 import Snackbar from '@material-ui/core/Snackbar';
 import MuiAlert, { AlertProps, Color } from '@material-ui/lab/Alert';
-
+import { AlertInterface } from '../../shared/types';
 import { makeStyles, Theme, createStyles } from '@material-ui/core/styles';
 
 export interface CustomSnackbarProps {
-  alert: { message: string; status: Color };
+  alert: AlertInterface;
   isOpen: boolean;
```
Comment (CedThen) [inline src/components/shared/Alert.tsx:10]: This was making my VS code angry so this was my fix

File: src/components/shared/NewScheduleForm.tsx [hunk 3]
```diff
@@ -124,11 +154,33 @@ export const NewScheduleForm = ({ onClose, error, onSubmit }: NewScheduleFormPro
...
+            templates.map(({ templateId, name }) => (
+              <MenuItem key={templateId} value={templateId}>
+                {name}
+              </MenuItem>
+            ))
+          ) : (
+            <>
+              <MenuItem value={1}>Weekly Services</MenuItem>
+              <MenuItem value={2}>RE</MenuItem>
+            </>
```
Comment (CedThen) [inline src/components/shared/NewScheduleForm.tsx:181]: This is placeholder, after shaun's merge, we can make the templates query just once

File: src/query/apis/schedules.ts [hunk 4]
```diff
@@ -17,21 +17,23 @@ export const getTabs = (churchId: number): Promise<AxiosResponse> => {
 export const getSchedule = (scheduleId: number): Promise<AxiosResponse> =>
   axios.get(`${secretIp}/api/schedules?scheduleId=${scheduleId}`);
 
-export const addSchedule = ({
+export const addSchedule = async ({
```
Comment (CedThen) [inline src/query/apis/schedules.ts:20]: typescript was griefing me here, will need to figure it out later :(

File: src/components/FormControl/ValidatedSelect.tsx [hunk 5]
```diff
@@ -13,6 +13,7 @@ interface ValidatedSelectProps {
   onChange: (arg: ValidatedFieldState<number>) => void;
   toolTip: { id: string; text: string };
   children: any[];
```
Comment (alan-mj-lin) [inline src/components/FormControl/ValidatedSelect.tsx:15]: proper typings

File: src/components/FormControl/ValidatedSelect.tsx [hunk 6]
```diff
@@ -13,6 +13,7 @@ interface ValidatedSelectProps {
...
+  label: string;
   [x: string]: any;
```
Comment (alan-mj-lin) [inline src/components/FormControl/ValidatedSelect.tsx:17]: proper typings

File: src/components/Templates/TemplateDisplay.tsx [hunk 7]
```diff
@@ -1,6 +1,94 @@
...
+import TableHead from '@material-ui/core/TableHead';
+import TableBody from '@material-ui/core/TableBody';
+import TableRow from '@material-ui/core/TableRow';
+import TableCell from '@material-ui/core/TableCell';
+import Button from '@material-ui/core/Button';
+
+import { makeStyles, Theme, createStyles } from '@material-ui/core/styles';
+
+interface TemplateDisplayProps {
+  template: any;
```
Comment (alan-mj-lin) [inline src/components/Templates/TemplateDisplay.tsx:18]: proper typings

File: src/components/Templates/Templates.tsx [hunk 8]
```diff
@@ -75,34 +36,82 @@ export const Templates = ({ churchId }: any) => {
...
+      cache.invalidateQueries('scheduleTabs');
+      closeDialogHandler(response);
+    },
+  });
+
+  function createNewTemplate() {}
 
-  if (isLoading) return <div>Loading</div>;
-  // console.log('data', templateData);
+  function closeDialogHandler(response?: any) {
```
Comment (alan-mj-lin) [inline src/components/Templates/Templates.tsx:49]: proper typings

File: src/components/shared/NewScheduleForm.tsx [hunk 9]
```diff
@@ -7,23 +7,35 @@ import { useValidatedField } from '../../hooks';
...
 import { buttonTheme } from '../../shared/styles/theme.js';
 
-import { Tooltip } from '../shared/Tooltip';
+import { Tooltip } from './Tooltip';
 import { stringLengthCheck } from '../../shared/utilities';
 // TODO hook up teams with data from DB
 
 interface NewScheduleFormProps {
   onClose: (data: any) => void;
   error: any;
```
Comment (alan-mj-lin) [inline src/components/shared/NewScheduleForm.tsx:16]: proper typings

File: src/components/shared/NewScheduleForm.tsx [hunk 10]
```diff
@@ -7,23 +7,35 @@ import { useValidatedField } from '../../hooks';
 
 import { buttonTheme } from '../../shared/styles/theme.js';
 
-import { Tooltip } from '../shared/Tooltip';
+import { Tooltip } from './Tooltip';
 import { stringLengthCheck } from '../../shared/utilities';
 // TODO hook up teams with data from DB
 
 interface NewScheduleFormProps {
   onClose: (data: any) => void;
```
Comment (alan-mj-lin) [inline src/components/shared/NewScheduleForm.tsx:15]: proper typings

File: src/components/shared/NewScheduleForm.tsx [hunk 11]
```diff
@@ -7,23 +7,35 @@ import { useValidatedField } from '../../hooks';
...
-  ) => void;
+  onSubmit: ({
+    scheduleTitle,
+    startDate,
+    endDate,
+    view,
+    team,
+    churchId,
+    templateId,
+  }) => Promise<any>;
```
Comment (alan-mj-lin) [inline src/components/shared/NewScheduleForm.tsx:25]: proper typings

File: src/components/shared/NewScheduleForm.tsx [hunk 12]
```diff
@@ -7,23 +7,35 @@ import { useValidatedField } from '../../hooks';
...
+  templateId?: number;
+  templates?: any;
```
Comment (alan-mj-lin) [inline src/components/shared/NewScheduleForm.tsx:27]: proper typings

File: server/utilities/helperFunctions.ts [hunk 13]
```diff
@@ -329,11 +330,25 @@ async function retrieveTaskData(eventId, role) {
...
       role: role,
     };
   });
   return organizedTasks;
 }
+
+export const recurringDaysOfWeek = (start: Date, end: Date, dayOfWeeK: number) => {
```
Comment (showduhtung) [inline server/utilities/helperFunctions.ts:342]: but.. but.. everyBleepDay..

File: src/query/apis/schedules.ts [hunk 14]
```diff
@@ -17,21 +17,23 @@ export const getTabs = (churchId: number): Promise<AxiosResponse> => {
...
   scheduleTitle,
   startDate,
   endDate,
   view,
   team,
   churchId,
-}: AddScheduleProps) =>
-  axios.post(`${secretIp}/api/schedules`, {
+  templateId,
+}: any) =>
```
Comment (alan-mj-lin) [inline src/query/apis/schedules.ts:28]: proper typings

File: src/components/Schedule/ScheduleContainer.tsx [hunk 15]
```diff
@@ -7,7 +7,7 @@ import { getTabData } from '../../query/schedules';
 import { addSchedule, deleteSchedule } from '../../query/apis/schedules';
 
 import { ScheduleTabs } from './ScheduleTabs';
-import { NewScheduleForm } from './NewScheduleForm';
+import { NewScheduleForm } from '../shared/NewScheduleForm';
```
Comment (showduhtung) [inline src/components/Schedule/ScheduleContainer.tsx:10]: This won't be a shareable form in the future, but for now it's fine as I've seen this was needed for POC

**Review by alan-mj-lin — CHANGES_REQUESTED** [review]
just a lot of typescript typing issues

File: src/components/shared/NewScheduleForm.tsx [hunk 16]
```diff
@@ -7,23 +7,35 @@ import { useValidatedField } from '../../hooks';
...
   error: any;
-  onSubmit: (
-    title: string,
-    startDate: string,
-    endDate: string,
-    view: string,
-    team: number,
-  ) => void;
+  onSubmit: ({
+    scheduleTitle,
```
Comment (showduhtung) [inline src/components/shared/NewScheduleForm.tsx:18]: I made changes in my branch about some of the Typescript stuff here and elsewhere. But we've done a lot of unnecessary destructuring. If you change this to:
```
onSubmit: (data: NewScheduleData) => Promise<any>;
```

it'll look a lot better. This data is also being passed around in a few components and functions, so the NewScheduleData interface is reusable. But I'll fix that in the merge conflict that'll definitely happen

File: src/components/shared/NewScheduleForm.tsx [hunk 17]
```diff
@@ -43,6 +55,12 @@ export const NewScheduleForm = ({ onClose, error, onSubmit }: NewScheduleFormPro
     0,
     'Please assign a team to this schedule',
   );
+  const [template, setTemplate, setTemplateError, resetTemplateError] = useValidatedField(
+    templateId ? templateId : 0,
```
Comment (showduhtung) [inline src/components/shared/NewScheduleForm.tsx:59]: templateId can be a number and string? I think this was when talking with @alan-mj-lin , but our ids are messed up all over the place, and inconsistent. From now on, let's assume id's are strings, unless we're comparing them

File: server/utilities/helperFunctions.ts (hunk 13 above)
Comment (CedThen) [reply server/utilities/helperFunctions.ts:342]: it doesn't do exactly the same thing haha
and i didn't know which function it became

File: scripts/database/templates.ts [hunk 18]
```diff
@@ -4,59 +4,138 @@ export const templates = [
...
-          { role: 'Interpreter', time: '11:00 AM' },
-          { role: 'Announcer', time: '12:00 PM' },
-          { role: 'Grace', time: '12:00 PM' },
+          { roleId: 1, title: 'Usher', time: '10:15 AM' },
+          { roleId: 2, title: 'Hymn Leading', time: '10:45 AM' },
+          { roleId: 3, title: 'Piano', time: '10:45 AM' },
+          { roleId: 1, title: 'Sermon Speaker', time: '11:00 AM' },
+          { roleId: 4, title: 'Interpreter', time: '11:00 AM' },
+          { roleId: 1, title: 'Announcer', time: '12:00 PM' },
+          { roleId: 6, title: 'Grace', time: '12:00 PM' },
```
Comment (hchiam) [inline scripts/database/templates.ts:27]: does it still work with `roleId: 1` repeated?

File: src/components/Templates/TemplateDisplay.tsx [hunk 19]
```diff
@@ -1,6 +1,94 @@
...
+                </TableRow>
+              ))}
+            </>
+          ))}
+        </TableBody>
+      </MaUTable>
+    </Card>
+  );
 };
+// howard u can work ur magic here hehe
```
Comment (hchiam) [inline src/components/Templates/TemplateDisplay.tsx:84]: noted :) might have time later today

File: src/query/apis/schedules.ts (hunk 14 above)
Comment (hchiam) [reply src/query/apis/schedules.ts:28]: maybe we can turn the linter rule back on later, after fixing the merge conflicts? 
we could default to requiring types, and only use "one-liner disablings" as needed, if we find we only really have to use `any` rarely

**Review by hchiam — CHANGES_REQUESTED** [review]

**Pull request merged**